orangecontrib/text/pubmed.py

```diff
@@ -0,0 +1,480 @@
+        """
+        if None in [self.search_record_web_env, self.search_record_query_key]:
+            raise ValueError(  # Perform a search first!
+                    'No WebEnv or QueryKey data in this PubMed class instance.'
+            )
+
+        fetch_handle = Entrez.efetch(
+                db='pubmed',
+                rettype='medline',
+                retmode='text',
+                retstart=batch_start,
+                retmax=batch_size,
+                webenv=self.search_record_web_env,
+                query_key=self.search_record_query_key
+        )
+
+        data = list(Medline.parse(fetch_handle))
+        fetch_handle.close()
+
+        with shelve.open(self.cache_path) as query_cache:
+            for record in data:
+                id_key = record.get('PMID')
+                if id_key is not None:
+                    query_cache[id_key] = record
+            query_cache.close()
+
+        if self.progress_callback:
+            self.progress_callback()
+
+        return data
+
+    def _retrieve_records(self, num_records,
+                          includes_metadata=PUBMED_TEXT_FIELDS,
+                          use_cache=True):
+        """Retrieves the records queried with '_search_for_records()'.
+
+        If retrieval was successful, generates a corpus with the text fields as
+        meta attributes.
+
+        Args:
+            num_records (int): The number of records we wish to retrieve.
+
+        Returns:
+            `orangecontrib.text.corpus.Corpus`: The retrieved PubMed records
+                as a corpus.
+        """
+        corpus = None
+        batch_size = min(self.DEFAULT_BATCH_SIZE, num_records)
+        cached_data = []  # Later on, construct the corpus from this.
+        new_records = []  # Must download.
+
+        if use_cache:
+            with shelve.open(self.cache_path) as query_cache:
+                # --- Retrieve cached ---
+                # case 1: [all cached]
+                # case 2: [all new]
+                # case 3: [cached + new]
+                for rec_id in self.record_id_list:
+                    if (len(cached_data) + len(new_records)) == num_records:
+                        break  # Capped.
+
+                    record = query_cache.get(rec_id)
+                    if record is not None:
+                        cached_data.append(record)
+                        continue
+
+                    new_records.append(rec_id)
+                query_cache.close()
+        else:
+            new_records = self.record_id_list
+
+        cached_data_size = len(cached_data)
+        if self.progress_callback:  # Advance the callback accordingly.
+            for _ in itertools.repeat(None, int(cached_data_size/batch_size)):
+                self.progress_callback()
+
+        # --- Retrieve missing/new ---
+        if len(new_records) > 0:
+            if cached_data_size > 0:  # Some records were cached.
```

Comment:
This is a copy of the code from the else below. By moving this if out of `if len(new_records) > 0:` we could eliminate one copy.